### PR TITLE
Skip shell functions when parsing env

### LINKED ISSUE #18: Gnome-terminal on Fedora 25
Installing this package for Atom on Fedora 25 and changing the terminal to 'gnome-terminal' has an unexpected side effect. It opens the terminal in the correct location how I would expect, it however prints a few lines in that terminal:

```bash: module: line 1: syntax error: unexpected end of file
bash: error importing function definition for `BASH_FUNC_module'
bash: scl: line 1: syntax error: unexpected end of file
bash: error importing function definition for `BASH_FUNC_scl'
/bin/sh: scl: line 1: syntax error: unexpected end of file
/bin/sh: error importing function definition for `BASH_FUNC_scl'
/bin/sh: module: line 1: syntax error: unexpected end of file
/bin/sh: error importing function definition for `BASH_FUNC_module'
/bin/sh: scl: line 1: syntax error: unexpected end of file
/bin/sh: error importing function definition for `BASH_FUNC_scl'
/bin/sh: module: line 1: syntax error: unexpected end of file
/bin/sh: error importing function definition for `BASH_FUNC_module'
/bin/sh: scl: line 1: syntax error: unexpected end of file
/bin/sh: error importing function definition for `BASH_FUNC_scl'
/bin/sh: module: line 1: syntax error: unexpected end of file
/bin/sh: error importing function definition for `BASH_FUNC_module'
[pjaleeuwen@localhost auth]$ 
```

### LINKED ISSUE #120: Strange lines in terminal
Hello !

In lastest Atom I have got this lines in terminal:
```
bash: module: line 1: syntax error: unexpected end of file
bash: error importing function definition for `BASH_FUNC_module'
bash: scl: line 1: syntax error: unexpected end of file
bash: error importing function definition for `BASH_FUNC_scl'
/bin/sh: scl: line 1: syntax error: unexpected end of file
/bin/sh: error importing function definition for `BASH_FUNC_scl'
/bin/sh: module: line 1: syntax error: unexpected end of file
/bin/sh: error importing function definition for `BASH_FUNC_module'
/bin/sh: scl: line 1: syntax error: unexpected end of file
/bin/sh: error importing function definition for `BASH_FUNC_scl'
/bin/sh: module: line 1: syntax error: unexpected end of file
/bin/sh: error importing function definition for `BASH_FUNC_module'
/bin/sh: scl: line 1: syntax error: unexpected end of file
/bin/sh: error importing function definition for `BASH_FUNC_scl'
/bin/sh: module: line 1: syntax error: unexpected end of file
/bin/sh: error importing function definition for `BASH_FUNC_module' 
```
What is this?
OS: Fedora 25

### LINKED ISSUE #165: Unexpected end of file sh: error importing function definition after update to v1.12.1
```
Error: sh: module: line 1: syntax error: unexpected end of file
sh: error importing function definition for `BASH_FUNC_module'
sh: scl: line 1: syntax error: unexpected end of file
sh: error importing function definition for `BASH_FUNC_scl'
    at parameters.exit (/home/msfernandes/.atom/packages/linter-flake8/node_modules/atom-linter/lib/helpers.js:47:27)
    at triggerExitCallback (/usr/share/atom/resources/app.asar/src/buffered-process.js:213:47)
    at /usr/share/atom/resources/app.asar/src/buffered-process.js:227:18
    at Socket.<anonymous> (/usr/share/atom/resources/app.asar/src/buffered-process.js:98:18)
    at emitOne (events.js:82:20)
    at Socket.emit (events.js:169:7)
    at Pipe._onclose (net.js:469:12)
```

### LINKED ISSUE #243: Error running Pylint Error: /bin/sh: module: line 1: syntax error: unexpected end of file
On Linux, pylint 1.8.2, Atom 1.24.
Python 3.4.6

No ~/.pylintrc present.
Fresh Atom installation:

```
[Linter] Error running Pylint Error: /bin/sh: module: line 1: syntax error: unexpected end of file
/bin/sh: error importing function definition for `module'
/bin/sh: mc: line 1: syntax error: unexpected end of file
/bin/sh: error importing function definition for `mc'
    at Object.<anonymous> (/homes/xxx/yyy/.atom/packages/linter-pylint/lib/main.js:162:17)
    at Generator.next (<anonymous>)
    at step (/homes/xxx/yyy/.atom/packages/linter-pylint/lib/main.js:7:273)
```

Any ideas? Thank you!

### LINKED ISSUE #293: getting errors while using terminal
when i load terminal it gives me below errors.
```
bash: error importing function definition for `scl'
/bin/sh: scl: line 1: syntax error: unexpected end of file
/bin/sh: error importing function definition for `scl'
/bin/sh: module: line 1: syntax error: unexpected end of file
/bin/sh: error importing function definition for `module'
/bin/sh: scl: line 1: syntax error: unexpected end of file
/bin/sh: error importing function definition for `scl'
/bin/sh: module: line 1: syntax error: unexpected end of file
/bin/sh: error importing function definition for `module'
/bin/sh: scl: line 1: syntax error: unexpected end of file
/bin/sh: error importing function definition for `scl'
/bin/sh: module: line 1: syntax error: unexpected end of file
/bin/sh: error importing function definition for `module'
```

OS : FEDORA 26
ATOM : 1.18.0-0.1

### LINKED ISSUE #347: some error when use autocomplete-python
### Steps to Reproduce
when i open atom, then some warning are showed such as below:
sh: module: line 1: syntax error: unexpected end of file
sh: error importing function definition for `module'
sh: scl: line 1: syntax error: unexpected end of file
sh: error importing function definition for `scl'

### LINKED ISSUE #422: flake8 keeps crashing again with BASH_FUNC_module
The error is "error importing function definition BASH_FUNC_module".

I used to have this problem but it's been gone for a while. Today I updated Atom from 1.15 to 1.16 and I possibly also updated linter-flake8, though I'm not sure about the latter. flake8 was at 3.2.1 and is now at 3.3.0 but the problem persists.

I had this problem a while back but upgrading flake8 fixed it. But now it has reappeared.

https://github.com/AtomLinter/linter-flake8/issues/165

### LINKED ISSUE #13451: Strange output
### Prerequisites

* [x] Can you reproduce the problem in [safe mode](http://flight-manual.atom.io/hacking-atom/sections/debugging/#using-safe-mode)?
* [x] Are you running the [latest version of Atom](http://flight-manual.atom.io/hacking-atom/sections/debugging/#update-to-the-latest-version)?
* [x] Did you check the [debugging guide](http://flight-manual.atom.io/hacking-atom/sections/debugging/)?
* [x] Did you check the [FAQs on Discuss](https://discuss.atom.io/c/faq)?
* [x] Are you reporting to the [correct repository](https://github.com/atom/atom/blob/master/CONTRIBUTING.md#atom-and-packages)?
* [x] Did you [perform a cursory search](https://github.com/issues?q=is%3Aissue+user%3Aatom+-repo%3Aatom%2Felectron) to see if your bug or enhancement is already reported?

### Description

Some packages that use comand-line/terminal gets a strange output.

### Steps to Reproduce

1. Get the latest atom version.
2. Install a plugin that uses the terminal, like git-control or platformio-atom-ide-terminal.
3. Open the terminal inside atom.

**Expected behavior:** A clean terminal; no pre-output.

**Actual behavior:** A strange output:
```
bash: module: line 1: syntax error: unexpected end of file
bash: error importing function definition for `BASH_FUNC_module'
bash: scl: line 1: syntax error: unexpected end of file
bash: error importing function definition for `BASH_FUNC_scl'
clear
/bin/sh: scl: line 1: syntax error: unexpected end of file
/bin/sh: error importing function definition for `BASH_FUNC_scl'
/bin/sh: module: line 1: syntax error: unexpected end of file
/bin/sh: error importing function definition for `BASH_FUNC_module'
/bin/sh: scl: line 1: syntax error: unexpected end of file
/bin/sh: error importing function definition for `BASH_FUNC_scl'
/bin/sh: module: line 1: syntax error: unexpected end of file
/bin/sh: error importing function definition for `BASH_FUNC_module'
/bin/sh: scl: line 1: syntax error: unexpected end of file
/bin/sh: error importing function definition for `BASH_FUNC_scl'

/bin/sh: module: line 1: syntax error: unexpected end of file
/bin/sh: error importing function definition for `BASH_FUNC_module'
```

### Versions

Atom    : 1.12.7
Electron: 1.3.13
Chrome  : 52.0.2743.82
Node    : 6.5.0
apm  1.12.9
npm  3.10.5
node 4.4.5
python 2.7.12
git 2.9.3
Fedora 25 (Linux)

### More info
I can't reproduce the problem in safe mode because atom doesn't initialize plugins.
See this: https://github.com/platformio/platformio-atom-ide-terminal/issues/120

### LINKED ISSUE #17369: atom-runner error
`/usr/bin/bash: module: line 1: syntax error: unexpected end of file
/usr/bin/bash: error importing function definition for `module'
/usr/bin/bash: switchml: line 1: syntax error: unexpected end of file
/usr/bin/bash: error importing function definition for `switchml'
/usr/bin/bash: scl: line 1: syntax error: unexpected end of file
/usr/bin/bash: error importing function definition for `scl'
Ignoring bindex-0.5.0 because its extensions are not built. Try: gem pristine bindex --version 0.5.0
Ignoring bootsnap-1.3.0 because its extensions are not built. Try: gem pristine bootsnap --version 1.3.0
Ignoring byebug-10.0.2 because its extensions are not built. Try: gem pristine byebug --version 10.0.2
Ignoring ffi-1.9.23 because its extensions are not built. Try: gem pristine ffi --version 1.9.23
Ignoring msgpack-1.2.4 because its extensions are not built. Try: gem pristine msgpack --version 1.2.4
Ignoring nio4r-2.3.1 because its extensions are not built. Try: gem pristine nio4r --version 2.3.1
Ignoring nokogiri-1.8.2 because its extensions are not built. Try: gem pristine nokogiri --version 1.8.2
Ignoring puma-3.11.4 because its extensions are not built. Try: gem pristine puma --version 3.11.4
Ignoring sqlite3-1.3.13 because its extensions are not built. Try: gem pristine sqlite3 --version 1.3.13
Ignoring websocket-driver-0.7.0 because its extensions are not built. Try: gem pristine websocket-driver --version 0.7.0`

first time using atom , but i got this when runmy first ruby code after ALT+R

### LINKED ISSUE #20389: process.env is corrupted on Fedora 31
### Prerequisites

* [X] Put an X between the brackets on this line if you have done all of the following:
    * Reproduced the problem in Safe Mode
    * Followed all applicable steps in the debugging guide:
    * Checked the FAQs on the message board for common solutions: <https://discuss.atom.io/c/faq>
    * Checked that your issue isn't already filed: 
(The issue has been filed before - for example issue #13451 - but they are closed as not reproducible in safe mode.  I have some more information, including how to reproduce in safe mode.)

### Description

The `process.env` contains garbage values when Atom is started on Fedora 31.  This seems to be related to [Environment Modules](http://modules.sourceforge.net/) which puts certain Bash functions with unusual names (such as `BASH_FUNC_switchml%%`)  into the environment - see `env1.txt` attachment, which contains the result of the `env` command before starting Atom.  My guess is whatever (Node?) code is initializing `process.env` is confused by this function, producing garbage - see attachment `process-env.txt`, which is from the debugger console.

This is almost certainly the cause of strange error message in terminals and other packages.  See issue #13451 

[process-env.txt](https://github.com/atom/atom/files/4172083/process-env.txt)

[env1.txt](https://github.com/atom/atom/files/4172006/env1.txt)

### Steps to Reproduce

1. If not running with Environment Modules (such as under Fedora 31), you can probably reproduce the problem by defining the `BASH_FUNC_switchml%%` functions as shown in the `env1.txt` attachment.
2. Start atom, optionally with `--safe`.
3. Open Developer Tools, and type `process.env` in the console.

(I would not surprised if the bug is easy to find by just looking at the relevant code, but I am not familiar with the Electron/Atom/Node code-base.)

**Expected behavior:**

A proper object, with one entry for each environment value.

**Actual behavior:**

Certain environment entries are spread out over multiple entries.  For example notice an extra entry for `" IFS"` (with initial space):

    process.env[' IFS']
    "$_mlIFS;"

Note I don't see this when starting DomTerm in a recent Electron, so this may have been fixed upstream, unless Atom does something weird.

**Reproduces how often:**

Always.
### Versions

    Atom    : 1.43.0
    Electron: 4.2.7
    Chrome  : 69.0.3497.128
    Node    : 10.11.0

Fedora 31 with Gnome/Wayland on x86_64.

### PR DESCRIPTION
**Pull request by xdelaruelle**
### Identify the Bug

Fixes #20389
Fixes #17369
Fixes #13451
Fixes blueimp/atom-open-terminal-here#27
Fixes blueimp/atom-open-terminal-here#18
Fixes bus-stop/Termination#101
Fixes bus-stop/terminus#24
Fixes platformio/platformio-atom-ide-terminal#120
Fixes platformio/platformio-atom-ide-terminal#293
Fixes AtomLinter/linter-pylint#243
Fixes AtomLinter/linter-flake8#643
Fixes AtomLinter/linter-flake8#165
Fixes AtomLinter/linter-flake8#422
Fixes AtomLinter/linter-puppet-lint#68
Fixes autocomplete-python/autocomplete-python#347

### Description of the Change

With this change `getEnvFromShell` now skips shell function definition to guarantee only environment variables are recorded and a sane `result` array is returned.

### Alternate Designs

Selected design fixes the issue with minimal change. But one may also question the need for `getEnvFromShell` function to execute in an external shell process the `env` command to get defined environment variables.

With this fix, the need for `ATOM_DISABLE_SHELLING_OUT_FOR_ENVIRONMENT` may also be questioned. Also `atom` binary could again be used directly by application link instead of `atom.sh` script.

### Possible Drawbacks

No drawback foreseen.

### Verification Process

Compile change in atom-dev, start terminus extensions: environment is found correctly defined and error messages disappeared.

### Release Notes

Skip shell functions when parsing env

### CODE REVIEW

**Comment by ESCHESLOW** [comment]
thank you

**Comment by TJProgrammer** [comment]
Thank you too!

**Comment by sadick254** [comment]
@xdelaruelle We appreciate your contribution.
This has the potential for affecting every atom package that calls an external program. The impact could be vast. Could you provide possible drawbacks you have thought about.

**Comment by xdelaruelle** [comment]
> @xdelaruelle We appreciate your contribution.
> This has the potential for affecting every atom package that calls an external program. The impact could be vast. Could you provide possible drawbacks you have thought about.

As you can see by the numerous references added to this issue, the impact is vast as it will solve a lot of existing issues found in many different atom packages.

Currently environment is parsed with the assumption that every line correspond to an environment variable. Which does not cope with environment functions that are spread over multiple lines. With this patch, functions are just ignored.

I do not foresee any drawback: environment functions are now ignored, but they were not intended to be found there and their parsing led to errors, so they should not miss to any package.

**Comment by xdelaruelle** [comment]
@sadick254 I have provided the information you require. Is there still something unclear to proceed with this pull request ?

**Comment by sadick254** [comment]
@xdelaruelle Thank you for all the info provided here.

**Review by sadick254 — APPROVED** [review]

File: src/update-process-env.js [hunk 1]
```diff
@@ -121,13 +121,22 @@ async function getEnvFromShell(env) {
   }
 
   let result = {};
+  let skip = false;
   for (let line of stdout.split('\n')) {
-    if (line.includes('=')) {
+    // start of shell function definition: skip full definition
+    if (line.includes('=() {')) {
```
Comment (ruv) [inline src/update-process-env.js:127]: What if a variable value contains the character sequence "=() {" ?
i.e. `x='a=() { b }'`

Comment (Tremont023323313) [reply src/update-process-env.js:127]: // end of shell function definition
(line==='}'){
skip=false;

**Review by ghost — COMMENTED** [review]
Thanks!

**Pull request merged**